lib/checksum.c

```diff
@@ -42,794 +41,858 @@
+    .free = (RmDigestFreeFunc)MurmurHash3_x86_128_free,
+    .update = (RmDigestUpdateFunc)MurmurHash3_x86_128_update,
+    .copy = (RmDigestCopyFunc)MurmurHash3_x86_128_copy,
+    .steal = (RmDigestStealFunc)MurmurHash3_x86_128_steal,
+#elif RM_PLATFORM_64
+    /* use 64-bit optimised murmur hash interface */
+    .new = (RmDigestNewFunc)MurmurHash3_x64_128_new,
+    .free = (RmDigestFreeFunc)MurmurHash3_x64_128_free,
+    .update = (RmDigestUpdateFunc)MurmurHash3_x64_128_update,
+    .copy = (RmDigestCopyFunc)MurmurHash3_x64_128_copy,
+    .steal = (RmDigestStealFunc)MurmurHash3_x64_128_steal,
+#else
+#error "Probably not a good idea to compile rmlint on 16bit."
+#endif
+};
+
+///////////////////////////
+//         metro         //
+///////////////////////////
+
+static const RmDigestInterface metro_interface = {
+    .name = "metro",
+    .bits = 128,
+    .len = NULL,
+    .new = (RmDigestNewFunc)metrohash128_1_new,
+    .free = (RmDigestFreeFunc)metrohash128_free,
+    .update = (RmDigestUpdateFunc)metrohash128_1_update,
+    .copy = (RmDigestCopyFunc)metrohash128_copy,
+    .steal = (RmDigestStealFunc)metrohash128_1_steal};
+
+static const RmDigestInterface metro256_interface = {
+    .name = "metro256",
+    .bits = 256,
+    .len = NULL,
+    .new = (RmDigestNewFunc)metrohash256_new,
+    .free = (RmDigestFreeFunc)metrohash256_free,
+    .update = (RmDigestUpdateFunc)metrohash256_update,
+    .copy = (RmDigestCopyFunc)metrohash256_copy,
+    .steal = (RmDigestStealFunc)metrohash256_steal};
+
+#if HAVE_SSE_4_2
+/* also define crc-optimised metro variants metrocrc and metrocrc256*/
+
+
+static const RmDigestInterface metrocrc_interface = {
+    .name = "metrocrc",
+    .bits = 128,
+    .len = NULL,
+    .new = (RmDigestNewFunc)metrohash128_1_new,  /* <-same */
+    .free = (RmDigestFreeFunc)metrohash128_free, /* <-same */
+    .update = (RmDigestUpdateFunc)metrohash128crc_update,
+    .copy = (RmDigestCopyFunc)metrohash128_copy, /* <-same */
+    .steal = (RmDigestStealFunc)metrohash128crc_1_steal};
+
+static const RmDigestInterface metrocrc256_interface = {
+    .name = "metrocrc256",
+    .bits = 256,
+    .len = NULL,
+    .new = (RmDigestNewFunc)metrohash256_new,    /* <-same */
+    .free = (RmDigestFreeFunc)metrohash256_free, /* <-same */
+    .update = (RmDigestUpdateFunc)metrohash256crc_update,
+    .copy = (RmDigestCopyFunc)metrohash256_copy, /* <-same */
+    .steal = (RmDigestStealFunc)metrohash256crc_steal};
 
-    g_cond_init(&self->change);
-    g_mutex_init(&self->lock);
-    return self;
+#endif
+
+///////////////////////////
+//      cumulative       //
+///////////////////////////
+
+#define RM_DIGEST_CUMULATIVE_MAX_BYTES 64
+
+#if RM_PLATFORM_64
+
+#define RM_DIGEST_CUMULATIVE_T guint64
+#define RM_DIGEST_CUMULATIVE_DATA data64
+#define RM_DIGEST_CUMULATIVE_ALIGN 8
+
+#else
+
+#define RM_DIGEST_CUMULATIVE_T guint32
+#define RM_DIGEST_CUMULATIVE_DATA data32
+#define RM_DIGEST_CUMULATIVE_ALIGN 4
+
```

Comment:
Did you try to run/compile that in 32bit envs? Otherwise I foresee that we'll get bug reports from those in about ~2 hours after a release. :)

Reply:
Ok I'll spin up a 32 bit VM and have a play

Reply:
Well that was fun (debian wheezy 32 bit).  Must say I learnt a few things.  Getting close to happy with the results but still have some bugs to fix.